log/format.go

```diff
@@ -44,188 +46,184 @@ type TerminalStringer interface {
 	TerminalString() string
 }
 
-func (h *TerminalHandler) TerminalFormat(r slog.Record, usecolor bool) []byte {
+func (h *TerminalHandler) TerminalFormat(buf []byte, r slog.Record, usecolor bool) []byte {
 	msg := escapeMessage(r.Message)
-	var color = 0
+	var color = ""
 	if usecolor {
 		switch r.Level {
 		case LevelCrit:
-			color = 35
+			color = "\x1b[35m"
 		case slog.LevelError:
-			color = 31
+			color = "\x1b[31m"
 		case slog.LevelWarn:
-			color = 33
+			color = "\x1b[33m"
 		case slog.LevelInfo:
-			color = 32
+			color = "\x1b[32m"
 		case slog.LevelDebug:
-			color = 36
+			color = "\x1b[36m"
 		case LevelTrace:
-			color = 34
+			color = "\x1b[34m"
 		}
 	}
+	if buf == nil {
+		buf = make([]byte, 0, 30+termMsgJust)
+	}
+	b := bytes.NewBuffer(buf)
 
-	b := &bytes.Buffer{}
-	lvl := LevelAlignedString(r.Level)
-	if color > 0 {
-		fmt.Fprintf(b, "\x1b[%dm%s\x1b[0m[%s] %s ", color, lvl, r.Time.Format(termTimeFormat), msg)
+	if color != "" { // Start color
+		b.WriteString(color)
+		b.WriteString(LevelAlignedString(r.Level))
+		b.WriteString("\x1b[0m")
 	} else {
-		fmt.Fprintf(b, "%s[%s] %s ", lvl, r.Time.Format(termTimeFormat), msg)
+		b.WriteString(LevelAlignedString(r.Level))
 	}
+	b.WriteString("[")
+	writeTimeTermFormat(b, r.Time)
+	b.WriteString("] ")
+	b.WriteString(msg)
+
 	// try to justify the log output for short messages
-	length := utf8.RuneCountInString(msg)
-	if r.NumAttrs() > 0 && length < termMsgJust {
-		b.Write(bytes.Repeat([]byte{' '}, termMsgJust-length))
+	//length := utf8.RuneCountInString(msg)
+	length := len(msg)
+	if (r.NumAttrs()+len(h.attrs)) > 0 && length < termMsgJust {
+		b.Write(spaces[:termMsgJust-length])
 	}
 	// print the keys logfmt style
 	h.logfmt(b, r, color)
 
 	return b.Bytes()
 }
 
-func (h *TerminalHandler) logfmt(buf *bytes.Buffer, r slog.Record, color int) {
-	attrs := []slog.Attr{}
-	r.Attrs(func(attr slog.Attr) bool {
-		attrs = append(attrs, attr)
-		return true
-	})
-
-	attrs = append(h.attrs, attrs...)
+func (h *TerminalHandler) logfmt(buf *bytes.Buffer, r slog.Record, color string) {
+	writeAttr := func(attr slog.Attr, first, last bool) {
+		buf.WriteByte(' ')
 
-	for i, attr := range attrs {
-		if i != 0 {
-			buf.WriteByte(' ')
+		if color != "" {
+			buf.WriteString(color)
+			buf.Write(appendEscapeString(buf.AvailableBuffer(), attr.Key))
+			buf.WriteString("\x1b[0m=")
+		} else {
+			buf.Write(appendEscapeString(buf.AvailableBuffer(), attr.Key))
+			buf.WriteByte('=')
 		}
-
-		key := escapeString(attr.Key)
-		rawVal := attr.Value.Any()
-		val := FormatLogfmtValue(rawVal, true)
+		tmp := buf.AvailableBuffer()
```

Comment:
`bytes.AvailableBuffer` was introduced in go1.21 .  can we make use of a separate temporary buffer, clearing it on each call to `appendEscapeString`?

Reply:
Good catch! Something temporary yes, lemme think about it...